### ISSUE TITLE
group and chord: subtasks are not executed until entire generator is consumed

### ISSUE THREAD

**Issue by chadrik** [celery/celery#3021]
When passing a generator to `group` or `chord`, sub-tasks are not submitted to the broker for execution until the entire generator is consumed.  The expected result is that subtasks are submitted during iteration, as soon as they are visited.

Here is the test code I'm using:

``` python
from celery import Celery, group
from celery.utils.log import get_task_logger

app = Celery(__name__, broker='redis://localhost:6379/0',
             backend='redis://localhost:6379/0')

logger = get_task_logger(__name__)

def longyield():
    import time
    for i in xrange(5):
        time.sleep(2)
        print "yielding", i
        yield add.s(i, i)

@app.task
def add(x, y):
    logger.info("adding {} + {}".format(x, y))
    return x + y

def itertest():
    it = longyield()
    g = group(it)
    result = g.delay()
    print result.get()
```

I'm testing with 4.0rc1.

**Comment by syradium** [comment]
Your example is really strange. Why do you need a group when you expect synchronous behaviour?

Well, I think you are wrong. Because first - you create a signature for a task which you call later. That task is a group which consists of `add` subtasks. This is created `yielding` from a `logyield`.

Later when the group is created and saved to `g` you call it a `add` subtasks are executed in parallel.

**Comment by chadrik** [comment]
@syradium Hopefully my example is not pointless because it is taken almost directly from the celery docs:

```python
from celery import group
res = group(add.s(i, i) for i in xrange(10))()
res.get(timeout=1)
```

I used `group.delay()` but that doesn't make a difference.  By extending the time it takes to yield results (e.g. `longyield` vs `xrange`) I'm trying to make apparent that `group` is consuming the entire generator before submitting subtasks for execution.

> Why do you need a group when you expect synchronous behaviour?

I don't understand your question.  I want all sub-tasks executed in parallel.  I thought that was the point of the `group` primitive. 

I'm also not the only person who has had this problem. I found [this old reference](https://groups.google.com/forum/#!topic/celery-users/9asJ1vCwmmE) from the celery-users google group, but the suggestions from Ask seemed out of date for the 3.1.x codebase, and the ones that I tried did not help.   I was surprised to find this issue has not been address in 4.x.

> ```python
> result = [add.delay(i, i).get() for i in longyield()]
> print result
> ```

Yes, that's the workaround I'm using for now, but the point is that I *want* to use the canvas primitives so that I can combine them into complex topologies.

**Comment by syradium** [comment]
@chadrik 

> @syradium Hopefully my example is not pointless because it is taken almost directly from the celery docs.

Yes and docs are correct. They are building a group and only then call it. Your example is a little bit different.

Look, this:
```
group(add.s(i, i) for i in xrange(10)
```

Is equivalent of:

```
group(add.s(0, 0), add.s(1, 1), ...)
```

Right? `add.s(0, 0)` does not start/execute a task. It only creates a subtask. You can later call it with `.delay()`/`apply_async()` or by running a group. So your tasks start running only after your generator stops.

**Comment by chadrik** [comment]
@syradium I got so annoyed trying to prove to you that you don't know what you're talking about that I just went ahead and fixed the problem.

The issue is that there's a test for empty tasks which forces the entire generator to be consumed before it's really needed (note the line `if not self.tasks`):

```
  File "/Users/chad/python/celery_walk_task.py", line 79, in itertest
    result = g()
  File "/Users/chad/dev/celery4.venv/lib/python2.7/site-packages/celery/canvas.py", line 807, in __call__
    return self.apply_async(partial_args, **options)
  File "/Users/chad/dev/celery4.venv/lib/python2.7/site-packages/celery/canvas.py", line 760, in apply_async
    if not self.tasks:
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/UserList.py", line 30, in __len__
    def __len__(self): return len(self.data)
  File "/Users/chad/dev/celery4.venv/lib/python2.7/site-packages/celery/utils/functional.py", line 353, in data
    self.__consumed.extend(list(self.__it))
  File "/Users/chad/python/celery_walk_task.py", line 77, in <genexpr>
    g = group(add.s(i, i) for i in longyield())
  File "/Users/chad/python/celery_walk_task.py", line 68, in longyield
    raise
```

The tasks are held in a `_regen` class, which caches the generator so that it can be consumed more than once, like a list (it is in fact a sublcass of `UserList` and `list`).  So, to get around the expansion of the generator just for this nonzero test, I added an optimized `__nonzero__` method:

```python
    def __nonzero__(self):
        if len(self.__consumed):
            return True
        try:
            # consume just one item
            self.__consumed.append(next(self.__it))
        except StopIteration:
            return False
        else:
            return True
```
By adding this, I was able to get past the nonzero test without consuming the entire generator, and the rest of the code handled the generator appropriately, submitting each task for async execution as it was yielded from the generator. 

I'm going to test this out a bit and submit a pull request.  Thanks for the motivation.

**Comment by syradium** [comment]
@chadrik  wow, take it easy.

I understand what are you talking about but wasn't sure we are on the same page. To me these two lines were identical (in celery terms):

```
group(add.s(i, i) for i in range(10))
group([add.s(i, i) for i in range(10)])
```

Be default I assumed that `celery` builds a list of whatever you give to it. Because in order to properly schedule a group of tasks it needs the whole group signature and not a generator which in fact can be infinite.

I do not see a use case for what you want to introduce. If each generator iteration is a long task so you want to start task in group as soon as possible, then why not to move this out to another task ...

Anyway, that's just my opinion. If community agree on that then ok.

**Comment by eli-green** [comment]
I think I understand the issue because I think it's one I've encountered. I'm a celery newbie so hopefully I won't mangle any terms too badly.

Let's say you have task A that generates chord B with callback C. A retrieves a few million rows from a database, does something to chunks of them in B and then merges the results in C. You're expecting the results from A to be very large but the results from B should be much smaller.

It would be ideal not to force A to keep its data around until B has been created. If there were a way for group/chord to start executing its tasks immediately, task A would only need to remember the GID of its subtasks (which could each be working on large chunks of data) instead of its entire data set. If A takes a long time to run, you could be losing the possibility of having a large amount of the work done by the time you finish generating B.

This would require some kind of call to mark B as "fully submitted for processing". From what I've seen of Redis as a broker (the only one I've used so far), it's using the native increment/decrement functions to check when the task is done. In this case you'd need to use those functions but also check that there wasn't potentially new work incoming for that chord. Just because I've completed 5/5 subtasks doesn't meant that A is done submitting tasks unless it says so.

It's certainly not trivial to implement.

**Comment by chadrik** [comment]
Yep, you've nailed it.  

I've explored this a bit more deeply and I've modified the celery code to work as desired for simple groups and it's almost there for chords.  

The way that chords in celery+redis work under the hood is this:

1. the number of sub-tasks in the group is computed and stored on the chord
2. the chord is stored on each group subtask
3. in the redis backend, when each subtask in the group completes, the number of completed subtasks is checked.  if the number complete equals the computed group size, then the chord callback is executed (except not quite. there's a little wrinkle which I'll explain later).

To put it another way, each group subtask carries a copy of the callback subtask forward, but only one -- the last to complete -- will execute it. 

And, of course, the first step described above, computing the number of tasks, necessarily means consuming the complete generator of subtasks, which is what I'm hoping to avoid.

The relevant backend code is here:

```python
    def on_chord_part_return(self, request, state, result, propagate=None):
        app = self.app
        tid, gid = request.id, request.group
        if not gid or not tid:
            return

        client = self.client
        jkey = self.get_key_for_group(gid, '.j')
        tkey = self.get_key_for_group(gid, '.t')
        result = self.encode_result(result, state)
        with client.pipeline() as pipe:
            _, readycount, totaldiff, _, _ = pipe                           \
                .rpush(jkey, self.encode([1, tid, state, result]))          \
                .llen(jkey)                                                 \
                .get(tkey)                                                  \
                .expire(jkey, 86400)                                        \
                .expire(tkey, 86400)                                        \
                .execute()

        totaldiff = int(totaldiff or 0)

        try:
            callback = maybe_signature(request.chord, app=app)
            total = callback['chord_size'] + totaldiff
            if readycount == total:
                decode, unpack = self.decode, self._unpack_chord_result
                with client.pipeline() as pipe:
                    resl, _, _ = pipe               \
                        .lrange(jkey, 0, total)     \
                        .delete(jkey)               \
                        .delete(tkey)               \
                        .execute()
                try:
                    callback.delay([unpack(tup, decode) for tup in resl])
                except Exception as exc:
                    error('Chord callback for %r raised: %r',
                          request.group, exc, exc_info=1)
                    return self.chord_error_from_stack(
                        callback,
                        ChordError('Callback error: {0!r}'.format(exc)),
                    )
        except ChordError as exc:
            error('Chord %r raised: %r', request.group, exc, exc_info=1)
            return self.chord_error_from_stack(callback, exc)
        except Exception as exc:
            error('Chord %r raised: %r', request.group, exc, exc_info=1)
            return self.chord_error_from_stack(
                callback,
                ChordError('Join error: {0!r}'.format(exc)),
            )
```

Looking at this, it doesn't seem that it would be that difficult to add a new group key along the lines of `.t` or `.j` which holds the total number of tasks, which is null (-1) until set by A when it submits the final task from group B.  

The only thing I don't understand yet is the `totaldiff` offset, e.g.

```python
            total = callback['chord_size'] + totaldiff
            if readycount == total:
                ...
```

Why can't we compare `readycount` directly to `callback['chord_size']`?  Why the need for the offset?  I'll have to do some more digging.

**Comment by chadrik** [comment]
Got it working without too much hassle.  I'll post a pull request soon.

**Comment by prydie** [comment]
This issue is currently causing me a lot of problems also.

I see that @chadrik's PR (#3043) has unfortunately not landed as of yet and he is no longer working on it at this point. I don't suppose any of the maintainers are able to pick up @chadrik's PR and get it ready for merging? I'm happy to help out with any testing, however, I'm not in a position to complete his work myself, unfortunately.

#### My use-case
I have three tasks:
 - `A`: Periodic task for the sending of weekly messages to customers
 - `B`: Task which does all the data crunching to generate each message and calls out to an external API to deliver the message
 - `C`: Callback for `chord` of task `B`s to report on sending job

To me this seems like an obvious use-case for `chord`, however, the amount of memory used by task `A` in generating the chord header is in the order of gigabytes.

**Comment by thedrow** [comment]
We'll try to get this on 4.1.

**Comment by RyanNaughton** [comment]
Hi, any update on this issue?

**Comment by thedrow** [comment]
No unfortunately none of us had time to work on this so far.

**Comment by thedrow** [comment]
For those who are still following this should hit 4.2 once we'll get a rebased version of #3043.
I'd be very happy for someone to volunteer to do so.

**Comment by thedrow** [comment]
This is finally fixed and will appear in 5.1.

**Comment by chadrik** [comment]
yay!

**Comment by RyanNaughton** [comment]
Awesome!

@pigrabbit87 Remember how this always used mess up our jobs! Flash from the past!

**Comment by thedrow** [comment]
I'm sorry to hear that it happened :) It must have been frustrating.
